it/src/main/java/com/google/cloud/teleport/it/pubsub/DefaultPubsubResourceManager.java

```diff
@@ -0,0 +1,220 @@
+import com.google.pubsub.v1.SubscriptionName;
+import com.google.pubsub.v1.Topic;
+import com.google.pubsub.v1.TopicName;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default class for implementation of {@link PubsubResourceManager} interface.
+ *
+ * <p>The class provides an interaction with the real Pub/Sub client, with operations related to
+ * management of topics and subscriptions.
+ */
+public final class DefaultPubsubResourceManager implements PubsubResourceManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultPubsubResourceManager.class);
+
+  private static final int DEFAULT_ACK_DEADLINE_SECONDS = 600;
+
+  private final String testId;
+  private final String projectId;
+  private final CredentialsProvider credentialsProvider;
+  private final TopicAdminClient topicAdminClient;
+  private final SubscriptionAdminClient subscriptionAdminClient;
+
+  private final List<TopicName> createdTopics;
+  private final List<SubscriptionName> createdSubscriptions;
+
+  private DefaultPubsubResourceManager(Builder builder) throws IOException {
+    this(builder.testId, builder.projectId, builder.credentialsProvider, null, null);
+  }
+
+  @VisibleForTesting
+  DefaultPubsubResourceManager(
+      @Nullable String testId,
+      String projectId,
+      CredentialsProvider credentialsProvider,
+      @Nullable TopicAdminClient topicAdminClient,
+      @Nullable SubscriptionAdminClient subscriptionAdminClient)
+      throws IOException {
+    this.projectId = projectId;
+    this.testId = (!Strings.isNullOrEmpty(testId) ? testId : "");
+    this.credentialsProvider = credentialsProvider;
+    this.topicAdminClient =
+        (topicAdminClient != null
+            ? topicAdminClient
+            : TopicAdminClient.create(
+                TopicAdminSettings.newBuilder()
+                    .setCredentialsProvider(credentialsProvider)
+                    .build()));
+    this.subscriptionAdminClient =
+        (subscriptionAdminClient != null
+            ? subscriptionAdminClient
+            : SubscriptionAdminClient.create(
+                SubscriptionAdminSettings.newBuilder()
+                    .setCredentialsProvider(credentialsProvider)
+                    .build()));
+    this.createdTopics = new ArrayList<>();
+    this.createdSubscriptions = new ArrayList<>();
+  }
+
+  public static Builder builder(String projectId) {
+    checkArgument(!projectId.isEmpty(), "projectId can not be empty");
+    return new Builder(projectId);
+  }
+
+  @Override
+  public TopicName createTopic(String topicName) {
+    LOG.info("Creating topic '{}'", topicName);
+
+    Topic topic = topicAdminClient.createTopic(getTopicName(topicName));
+    TopicName reference = PubsubUtils.toTopicName(topic);
+    createdTopics.add(reference);
```

Comment:
To make sure that this class is thread safe, we should either use a thread-safe collection or make these calls in a `synchronize` block. (I think synchronizing the whole method is a bit much.)

Reply:
Thanks. Added `Collections.synchronizedList` above.